code/variance_test.py

```
@@ -140,15 +159,41 @@ def __delta(self, X, q: int, j: int):
         numerator = float(np.dot(lead ** 2, lagged ** 2))
         return (upper_bound * numerator) / (denominator ** 2)
 
+    @staticmethod
+    def __delta_from_diffs(
+        diffs: np.ndarray, denominator: float, upper_bound: int, j: int
+    ) -> float:
+        """Compute the delta statistic using precomputed differences."""
+
+        if j <= 0 or j >= upper_bound:
+            raise ValueError("Lag j must satisfy 0 < j < upper_bound.")
```

Ruff (0.14.6)

Avoid specifying long messages outside the exception class

(TRY003)